File: 00003_longest_substr/solution.go
func lengthOfLongestSubstring(s string) int {
    i, j := 0, 0
    rs := []rune(s)
    
    h := make(map[rune]int)
    buffer := []rune("")
    winner := []rune("")
    
    for j < len(rs) {
        c := rs[j]
        if !containsRune(buffer, c) {
            buffer = append(buffer, c)
            h[c] = j
            j += 1
        } else {
            i = h[c] + 1
            if len(buffer) > len(winner) {
                winner = buffer
            }
            buffer = rs[i:j]
            h[c] = j
        }
    }
    
    if len(buffer) > len(winner) {
        winner = buffer
    }
    
    return len(winner)
}

func containsRune(rs []rune, r rune) bool {
    for i := 0; i < len(rs); i++ {
        if r == rs[i] {
            return true
        }
    }
    return false
}

File: 00003_longest_substr/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        currWinner = ""
        buffer = ""
        h = {}
        i, j = 0, 0
        while j < len(s):
            currChar = s[j]
            if s[j] not in buffer:
                h[s[j]] = j
                buffer += s[j]
                j += 1
                
            else:
                i = h[s[j]] + 1
                if len(buffer) > len(currWinner):
                    currWinner = buffer
                    
                buffer = s[i:j]
                h[s[j]] = j
                
        if len(buffer) > len(currWinner):
            currWinner = buffer
                
        return len(currWinner)
